src/calendar-app/calendar/model/CalendarModel.ts

```diff
@@ -200,19 +212,161 @@ export class CalendarModel {
+		let calendarGroupRootsList: CalendarGroupRoot[] = []
+		for (const membership of userController.getCalendarMemberships()) {
+			groupRootsPromises.push(this.entityClient.load(CalendarGroupRootTypeRef, membership.group))
+		}
+		calendarGroupRootsList = await Promise.all(groupRootsPromises)
+
+		if (!existingGroupSettings) {
+			const { groupSettings: gSettings } = await locator.entityClient.load(UserSettingsGroupRootTypeRef, userController.user.userGroup.group)
+			existingGroupSettings = gSettings
+		}
+
+		for (const { sourceUrl, group } of existingGroupSettings) {
+			if (!sourceUrl) {
+				continue
+			}
+
+			const lastSync = this.deviceConfig.getLastExternalCalendarSync().get(group)
+			const offset = 1000 // Add an offset to account for cpu speed when storing or generating timestamps
+			if (lastSync && Date.now() + offset - lastSync < syncInterval) {
+				continue
+			}
+
+			const currentCalendarGroupRoot = calendarGroupRootsList.find((calendarGroupRoot) => isSameId(calendarGroupRoot._id, group)) ?? null
+			if (!currentCalendarGroupRoot) {
+				console.error(`Trying to sync a calendar the user isn't subscribed to anymore: ${group}`)
+				continue
+			}
+
+			const externalCalendar = await this.fetchExternalCalendar(sourceUrl)
+			const parsedExternalEvents = parseCalendarStringData(externalCalendar, getTimeZone()).contents
+
+			const existingEventList = await loadAllEvents(currentCalendarGroupRoot)
+
+			const operationsLog: {
+				skipped: CalendarEvent[]
+				updated: CalendarEvent[]
+				created: CalendarEvent[]
+				deleted: CalendarEvent[]
+			} = {
+				skipped: [],
+				updated: [],
+				created: [],
+				deleted: [],
+			}
+			/**
+			 * Sync strategy
+			 * - Replace duplicates
+			 * - Add new
+			 * - Remove rest
+			 */
+			const { rejectedEvents, eventsForCreation } = sortOutParsedEvents(parsedExternalEvents, existingEventList, currentCalendarGroupRoot, getTimeZone())
+			const duplicates = rejectedEvents.get(EventImportRejectionReason.Duplicate) ?? []
+
+			// Replacing duplicates with changes
+			for (const duplicatedEvent of duplicates) {
+				const existingEvent = existingEventList.find((event) => event.uid === duplicatedEvent.uid)
+				if (!existingEvent) {
+					console.warn("Found a duplicate without an existing event!")
+					continue
+				}
+				if (this.eventHasSameFields(duplicatedEvent, existingEvent)) {
+					operationsLog.skipped.push(duplicatedEvent)
+					continue
+				}
+				await this.updateEventWithExternal(existingEvent, duplicatedEvent)
+				operationsLog.updated.push(duplicatedEvent)
+			}
+			console.log(TAG, `${operationsLog.skipped.length} events skipped (duplication without changes)`)
+			console.log(TAG, `${operationsLog.updated.length} events updated (duplication with changes)`)
+
+			// Add new event
+			for (const { event } of eventsForCreation) {
+				const { assignEventId } = await import("../../../common/calendar/date/CalendarUtils")
```

Comment:
Do we need to await the import in the loop? Or could it be done before?

Reply:
Good catch I missed this